### ISSUE TITLE
Unable to decrypt SSL in Wireshark if captured in "regular" mode. Works in "transparent" mode.

### ISSUE THREAD

**Issue by idiotsavant** [mitmproxy/mitmproxy#456]
I am using 0.11.3 on CentOS 6.6. I run the proxy in transparent mode and use iptables to redirect traffic as detailed in the tutorial. The mitdump command is:

`mitmdump -T -p 7777 -s 'responseMod.py' --anticache --ciphers '!DH:!ECDH:HIGH:MEDIUM:!aNULL:!MD5'`

The traffic flow is:

`[Firefox on VM host] <--> [eth1 on VM running proxy]-[eth0 on VM running proxy] <--> [Website]`

This works fine; the responses of HTTPS pages are modified as expected.

I am simultaneously capturing the traffic using tcpdump on eth1:

`tcpdump -ni eth1 -s0 port 443 -w /tmp/transparent.pcap`

Again, this works fine. I can use the mitmproxy-ca.pem key in Wireshark and decrypt the traffic.

The problem is if I switch from transparent to regular mode. I remove the -T option in mitmdump and the iptables rules, and configure Firefox to use the proxy explicitly on 7777 (all protocols) and run:

`mitmdump -p 7777 -s 'responseMod.py' --anticache --ciphers '!DH:!ECDH:HIGH:MEDIUM:!aNULL:!MD5'`

Now, if I capture in tcpdump I have to specify port 7777:

`tcpdump -ni eth1 -s0 port 7777 -w /tmp/regular.pcap`

But I cannot decrypt the traffic in Wireshark. Everything is fine in Firefox in both regular and transparent mode - the response of HTTPS pages is modified as expected - but I need the ability to passively capture and decrypt the traffic at the same time, and that only works in transparent mode.

I should say up front that I don't know if this is a bug, a limitation, or just a reflection of my lack of deep understanding of SSL decryption and how it works. Is this expected behaviour? I want to use regular mode if possible as it makes it much easier for users to set up.

Many thanks!

**Comment by mhils** [comment]
Hi @idiotsavant,

> But I cannot decrypt the traffic in Wireshark. Everything is fine in Firefox in both regular and transparent mode - the response of HTTPS pages is modified as expected - but I need the ability to passively capture and decrypt the traffic at the same time, and that only works in transparent mode.

This terribely sounds like a bug in Wireshark. In fact, I can even reproduce it :smile: 
I just recorded a small conversation on my Ubuntu VM - once I select "Follow SSL stream" an empty window pops up. However, if I take a look at the Wireshark SSL debug file (SSL Protocol Preferences), I see multiple snippets similar to this
```
ssl_decrypt_record: allocating 4368 bytes for decrypt data (old len 624)
Plaintext[4336]:
| 04 57 f6 35 94 6f a6 93 2d 5f 20 52 ab ff c4 17 |.W.5.o..-_ R....|
| 48 54 54 50 2f 31 2e 31 20 32 30 30 20 4f 4b 0d |HTTP/1.1 200 OK.|
| 0a 53 65 72 76 65 72 3a 20 6e 67 69 6e 78 2f 31 |.Server: nginx/1|
| 2e 31 2e 31 39 0d 0a 44 61 74 65 3a 20 4d 6f 6e |.1.19..Date: Mon|
<snip>

ssl_decrypt_record found padding 11 final len 4324
checking mac (len 4288, version 303, ct 23 seq 1)
tls_check_mac mac type:SHA1 md 2
Mac[20]:
| 15 ab 52 e5 06 12 de a1 05 e0 d4 fa 1d 6e 33 1a |..R..........n3.|
| 55 3a 59 46                                     |U:YF            |
ssl_decrypt_record: mac failed
```
As it seems, decryption works, but there is some MAC error which makes Wireshark fail. I copied my `mitmproxy-ca.pem` and the pcap file to my windows machine - decryption works perfectly here.

TL;DR: You should get a Windows box :stuck_out_tongue: 

Cheers,
Max

**Comment by idiotsavant** [comment]
Hi Max,

I saw the same thing when I had the proxy in transparent mode: I was unable to see my decrypted traffic in the Wireshark GUI but I could see the plaintext in the debug file. I didn't mention that it wasn't working for me in the Wireshark GUI because I wanted to keep the explanation of the symptoms simple.

If I capture traffic with the proxy regular mode I do NOT see any plain text in the debug file.

This isn't just a Wireshark bug, because my test with tcpdump and Wireshark is just an illustration of the problem using tools that are widely available. What I'm actually using is a passive capture application that captures in a way very similar to tcpdump, and using that I see the same problem: I can decrypt captured SSL traffic only if the proxy is in transparent mode.

Many thanks!

Dan

**Comment by mhils** [comment]
Ok, then I can't reproduce it - apart from the problems mentioned above, it works fine for me on Windows and Linux in regular mode. What's stated in the Wireshark debug log? Which cipher is negotatied in the handshake? We defer all SSL processing to (py)OpenSSL, there's not really much we can do wrong if the connection works.

Cheers,
Max

**Comment by idiotsavant** [comment]
I'll gather that information together for you shortly. Thanks for taking a look!

One question: do you have two interfaces on the box where mitmproxy is running? I have it on a VM in VirtualBox with eth1 on a host-only network and eth0 connected to my router using DHCP. I'm wondering if that might be a contributing factor.

Cheers,

Dan

**Comment by idiotsavant** [comment]
Hi Max,

I'm linking to two small .pcap files, captured as detailed in my first post, and the corresponding Wireshark debug files. The larger one has lots of plaintext content and corresponds to the traffic captured whilst the proxy was in transparent mode.

The cipher in both cases was TLS_RSA_WITH_AES_128_CBC_SHA (0x002f).

Cheers,

Dan

https://dl.dropboxusercontent.com/u/18927567/2xPCAPand2xWiresharkLogs.zip (5 MB)

**Comment by idiotsavant** [comment]
OK, my mistake!
I hadn't updated the port in the RSA keys window in Wireshark to change it from 443 to 7777. I now see plain text in with the traffic captured in the regular proxy mode. Which means the issue is with my capture app, and not Wireshark or mitmproxy.

Many thanks for your help, and sorry to waste your time.

Dan

**Comment by mhils** [comment]
No problem. Thanks for clarifying!

Cheers,
Max
